### PR TITLE
Added Less-Forgetful Learning LFL Plugin

### PR DESCRIPTION
**Pull request by Ektagavas**
I have implemented the LFL plugin from the paper [Less-Forgetful Learning](https://arxiv.org/abs/1607.00122) along with the strategy wrapper, and an example.

The LFL algorithm requires computing penalty euclidean loss with current and previous model features. To avoid model-specific modifications in the plugin (i.e flattening the input in case of MLP model), I have created a base abstract class `BaseModel `which requires implementing `get_features()` in its subclasses, as suggested by @AndreaCossu on Slack. 

Currently, I have only made the `SimpleMLP` class inherit this `BaseModel` for the example and test cases to work. The plugin will check if the model is an instance of this base class and will raise NotImplementedError if it is not. 

Kindly let me know if the implementation seems okay.
Thank you! :)

### CODE REVIEW

File: avalanche/training/plugins/lfl.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import torch
+
+from avalanche.training.plugins.strategy_plugin import StrategyPlugin
+from avalanche.training.utils import get_last_fc_layer, freeze_everything
+from avalanche.models.base_model import BaseModel
+
+
+class LFLPlugin(StrategyPlugin):
+
+    def __init__(self, lambda_e):
```
Comment (AntonioCarta) [inline avalanche/training/plugins/lfl.py:12]: can you add more details here? or at least link to `LFL(BaseStrategy)`'s documentation.

File: avalanche/training/strategies/strategy_wrappers.py [hunk 2]
```diff
@@ -618,6 +619,52 @@ def __init__(self, model: Module, optimizer: Optimizer, criterion,
...
 
+class LFL(BaseStrategy):
+
+    def __init__(self, model: Module, optimizer: Optimizer, criterion,
+                 lambda_e: Union[float, Sequence[float]],
+                 train_mb_size: int = 1, train_epochs: int = 1,
+                 eval_mb_size: int = None, device=None,
+                 plugins: Optional[List[StrategyPlugin]] = None,
+                 evaluator: EvaluationPlugin = default_logger, eval_every=-1):
+        """ Less Forgetful Learning strategy.
```
Comment (AntonioCarta) [inline avalanche/training/strategies/strategy_wrappers.py:630]: can you add a link to the paper?

**Comment by AntonioCarta** [comment]
Thanks @Ektagavas! 
Are you able to reproduce the paper's results?

File: avalanche/training/strategies/strategy_wrappers.py (hunk 2 above)
Comment (Ektagavas) [reply avalanche/training/strategies/strategy_wrappers.py:630]: Yes. I will update it. Thanks!

**Comment by AntonioCarta** [comment]
> The paper reports results on domain adaptation scenarios. I am currently running a few experiments and will get back soon on this.

You should be able to use avalanche for domain adaptation experimentation. If it's not too difficult, having a minimal example reproducing published results helps us a lot.

> can someone please help me find out how the unit tests are failing now?

All the tests have passed successfully, maybe the problem was in a previous commit.

**Comment by AndreaCossu** [comment]
Hi @Ektagavas , the test errors were caused by temporary errors in the HTTP requests. I re-run them and it all worked fine.

**Comment by Ektagavas** [comment]
> Hi @Ektagavas , the test errors were caused by temporary errors in the HTTP requests. I re-run them and it all worked fine.

Hi, that's great. Thanks.

> > The paper reports results on domain adaptation scenarios. I am currently running a few experiments and will get back soon on this.
> 
> You should be able to use avalanche for domain adaptation experimentation. If it's not too difficult, having a minimal example reproducing published results helps us a lot.
> 
> > can someone please help me find out how the unit tests are failing now?
> 
> All the tests have passed successfully, maybe the problem was in a previous commit.



> Thanks @Ektagavas!
> Are you able to reproduce the paper's results?


I carried out following two basic sets of experiments using the LFL plugin with Avalanche, similar to what the authors had considered  and reported in Table 5 in their domain-expansion [paper ](https://arxiv.org/pdf/1711.05959.pdf)

'Old' columns refer to test accuracy on respective datasets after training on D1 and similarly, 'New' refers to accuracy after training on D2

1) MNIST(D1) -> SVHN(D2). I used resnet18 model and trained for 2 epochs.

|   MNIST(D1) -> SVHN(D2)  | (Old) <br>D1 (%) | (Old) <br>D2 (%) | (New) <br>D1 (%)| (New) <br>D2 (%)| Avg(%)| Forgetting  (%) |
| ---------------------- | ----------------------- | ---------------------- |--- | ----------------------- | ---------------------- | ----- |
| Naive                  | 98.68                   | 20.96                  | 74.35 | 90.01 | 82.18 | 24.33 |
| EWC(ewc\_lambda=0.001) | 98.69                   | 19.67                  | 66.89 | 89.51 | 78.2 | 31.8 |
| EWC(ewc\_lambda=0.01)  | 98.4                    | 19.65                  | 72.39 | 89.01 | 80.7 | 26.01 |
| EWC(ewc\_lambda=0.1)   | 98.55                   | 20.23                  | 77.72 | 88.21 | 82.96 | 20.83 |
| LFL(lambda\_e=0.01)    | 98.61                   | 20.18                  | 94.43 | 84.76 | 89.6 | 4.18 |
| LFL (lambda\_e=0.1)    | 98.94                   | 21.12                  | 94.03 | 84.27 | 89.15 | 4.91 |
| LFL (lambda\_e=1)      | 98.61                   | 20.18                  | 96.64 | 84.67 | <b>90.66</b> | 1.97 |


2) CIFAR10 Color (D1)-> CIFAR10 Gray (D2). Here, model is resnet18 and epochs=10

| CIFAR10 Color (D1) -> Gray (D2)| (Old) <br>D1 (%) | (Old) <br>D2 (%) | (New) <br>D1 (%)| (New) <br>D2 (%)| Avg (%)| Forgetting  (%) |
| ---------------------- | ------------- | ------------ | ------------- | ----- | ------- | ---------- |
| Naive                  | 71.43 | 63.49 | 72.16 | 69.24 | 70.7    | \-0.73     |
| EWC(ewc\_lambda=0.001) | 72.45 | 64.86 | 71.91 | 69.68 | 70.8    | 0.54       |
| EWC(ewc\_lambda=0.01)  | 72.81 | 64.67 | 71.73 | 69.77 | 70.75   | 1.08       |
| EWC(ewc\_lambda=0.1)   | 71.51 | 64.44 | 71.9  | 69.71 | 70.81   | \-0.39     |
| LFL(lambda\_e=0.01)    | 72.06 | 64.28 | 73.59 | 70.07 | 71.83   | \-1.53     |
| LFL (lambda\_e=0.1)    | 73.28 | 66    | 73.85 | 70.18 | <b>72.02</b>   | \-0.57     |
| LFL (lambda\_e=1)      | 71.43 | 63.49 | 72.9  | 68.25 | 70.58   | \-1.47     |

Overall, the above values show that LFL outperforms Naive and EWC as consistent with paper results. 

Kindly let me know your views. Thanks!

**Comment by AntonioCarta** [comment]
Thanks @Ektagavas, everything looks good. If you want, you can open a new PR with your scripts to reproduce the paper's results, surely they can be helpful for other people working with LFL.

**Comment by Ektagavas** [comment]
> Thanks @Ektagavas, everything looks good. If you want, you can open a new PR with your scripts to reproduce the paper's results, surely they can be helpful for other people working with LFL.

Thank you!
Sure, I'll open a new PR with my scripts to reproduce the above experiments. Should I place the scripts in 'examples' folder?
